### ISSUE TITLE
Performance when serializing Proxy objects

### ISSUE THREAD

**Issue by Elberet** [larshp/abapGit#2860]
Our application communicates with dozens of webservices in many supported versions; for each webservice, each version and each method our system contains at least one SPRX object.

When serializing SPRX objects, AbapGit loads the proxy object database once *for each serialized object*:
https://github.com/larshp/abapGit/blob/971efbfaac4ccd3689636d5d9fba342b97acbe6e/src/objects/zcl_abapgit_object_sprx.clas.abap#L181-L185

This is an expensive operation involving at least a linear search over a database table of ~350k entries returning on the order of 2000 records. All in all, each proxy object takes several seconds to serialize only because of this; and since the operation is already IO-bound, performance does not scale with parallel processing.

In fact, when parallel processing is enabled, this results in AbapGit crashing with an assertion error in [zcl_abapgit_serialize.clas.abap#L199](https://github.com/larshp/abapGit/blob/5cd20447450be0668aebe177557cbdd63b4153d7/src/zcl_abapgit_serialize.clas.abap#L199).

Is it possible to improve performance here? Perhaps cache some of the loaded records between serializations of different SPRX objects?

**Comment by larshp** [comment]
what happens if you set `iv_force_sequential`

I'm also interested in why it dumps, it should run regardless of what happens in the thread

**Comment by larshp** [comment]
looking at the comment in the code, `cl_proxy_db=>load_by_abap_name` does not exist in lower releases, so we can add a dynamic call to this one, and fallback to `load`, guess it will solve the issue for you?

**Comment by Elberet** [comment]
In sequential mode, serialization doesn't dump but performance remains abysmally slow - no surprise there. As to why it dumps in parallel mode, the [serialize loop](https://github.com/larshp/abapGit/blob/5cd20447450be0668aebe177557cbdd63b4153d7/src/zcl_abapgit_serialize.clas.abap#L276-L293) looks broken. If no thread returns within 10 seconds, the wait statement returns while `mv_free = 0` and the next call to `run_parallel` fails the `mv_free > 0` assertion.

I haven't looked into `cl_proxy_db=>load_by_abap_name`'s internals yet, so until I get back to the office on Monday I can't say if that'd fix the performance, and I don't have sufficient experience with the metal representation of proxy objects to hazard a guess right now. 😉

**Comment by Elberet** [comment]
Apparently `zcl_abapgit_object_sprx->load_db` is a straight reimplementation of `cl_proxy_db=>load_by_abap_name`, at least I can't see any significant differences. As such, I doubt that calling the latter will affect performance at all.

Unfortunately, I can't invest any more time into this. 😞

**Comment by larshp** [comment]
Performance trace for test repo https://github.com/abapGit-tests/SPRX looks okay, OP does not have more time to invest, closing, unable to reproduce

![image](https://user-images.githubusercontent.com/5888506/66294632-b8fb1f00-e8e9-11e9-98e9-e9653eafb0ee.png)

**Comment by Elberet** [comment]
Hey @larshp, just a final observation before this issue is consigned to oblivion:

The performance test shows that a total of ~5ms is spent reading proxy object infos from the database. I believe that this scales _quadratically_ with the number of proxy objects being serialized, because each proxy object triggers a search in the SPROXHDR table and each object also makes searching that table more expensive.

I'd be really curious what the performance stats would look like for 10, 100 and 1000 proxy objects in the repository... 😅

**Comment by larshp** [comment]
It could be nice to look into it, but need more information, what is the exact SELECT statement that takes up the time?

**Comment by Elberet** [comment]
It's actually not even the headers table, but the table `SPROXDAT` (with 11ms reads in your performance trace). This appears to be the expensive statement:

![image](https://user-images.githubusercontent.com/361662/66387857-262fb280-e9c5-11e9-9d79-2c82d4157a13.png)

In one execution, the itab `sproxhdr_tab` contained 405 entries.
In `zcl_abapgit_object_sprx=>load_db`, `zcl_proxy_db=>load` was called with `lt_ids` and `lt_packages` each containing 1 entry and `ld_hdr-gen_appl = 'WEBSERVICES'`.

The DB table `SPROXDAT` contains ~384000 entries.

The SQL trace shows that a single select statement can take up to 8 seconds:
![image](https://user-images.githubusercontent.com/361662/66390417-eff53180-e9ca-11e9-9db2-f4afdc8d217b.png)

**Comment by larshp** [comment]
Looks like a database problem to me, it _should_ find the values via partial primary key, try doing an explain on the statements

Also try installing SAP note 2628942, dont think the note will make a difference, but it is worth a try.

**Comment by Elberet** [comment]
The query plan looks harmless enough.
I'll ping our SAP basis guys with the node and try again later...

**Comment by Elberet** [comment]
Good news, installing the note you've mentioned seems to have helped _a lot_. The sheer number of queries is way down; largest I've seen so far is on the order of 80 identical queries on `SPROXDAT`, where before I had seen a single query up to 260 times. Query times are still not stellar with ~8ms per individual query, but it's a busy system.

Bad news, serializing a single proxy object may, occasionally, still take >10 seconds. As mentioned in the OP, this triggers a bad assumption in the parallel processing code and silently crashes abapGit on an `ASSERT` statement because `run_parallel` may be called even when `mv_free = 0` [here](https://github.com/larshp/abapGit/blob/5cd20447450be0668aebe177557cbdd63b4153d7/src/zcl_abapgit_serialize.clas.abap#L287). Should I open a separate issue for this?

**Comment by larshp** [comment]
update to latest abapGit version, the timeout has been changed to 120sec, https://github.com/larshp/abapGit/blob/master/src/zcl_abapgit_serialize.clas.abap#L301
in commit
https://github.com/larshp/abapGit/commit/cb341e778fe3ced8c15ff2b8a317228556cc9d52
